File: javascript/packages/base/src/AttestationValidationContext.js
/**
 * AttestationValidationContext
 *
 * Provides validation context for the attestation validation pipeline,
 * including cycle detection via a seen set, depth tracking, and a reference
 * to the validation function for recursive validation.
 *
 * This context is shared across all stages and specialists in the validation pipeline,
 * allowing them to:
 * - Record visited attestation UIDs (detect cycles)
 * - Track recursion depth (prevent unbounded chains)
 * - Access the Merkle root from the parent document
 * - Call back into the pipeline for recursive validation
 */

/**
 * Creates an attestation validation context.
 *
 * @param {Object} options - Configuration options
 * @param {string} [options.merkleRoot] - Optional Merkle root from the parent document
 * @param {Record<string, unknown>} [options.extension] - Optional extension bag for custom data
 * @param {number} [options.maxDepth=32] - Maximum recursion depth
 * @returns {Object} The validation context
 * @throws {Error} If maxDepth is invalid
 */
export function createAttestationValidationContext(options = {}) {
    const {
        merkleRoot = null,
        extension = {},
        maxDepth = 32
    } = options;

    if (!Number.isInteger(maxDepth) || maxDepth < 1) {
        throw new Error('maxDepth must be a positive integer');
    }

    // Seen set to track visited UIDs for cycle detection
    const seen = new Set();

    // Current recursion depth
    let currentDepth = 0;

    // validateAsync will be set by the pipeline when context is created
    let validateAsync = null;

    return {
        // Merkle root from parent document
        merkleRoot,

        // Extension bag for custom data
        extension,

        // Maximum allowed recursion depth
        maxDepth,

        /**
         * Records a visit to an attestation UID.
         * Throws if the UID has already been visited (cycle detection).
         *
         * @param {string} attestationUid - The UID to record
         * @throws {Error} If the UID has already been visited
         */
        recordVisit(attestationUid) {
            if (!attestationUid || typeof attestationUid !== 'string') {
                throw new Error('attestationUid must be a non-empty string');
            }

            if (seen.has(attestationUid)) {
                throw new Error(`Cycle detected: attestation UID ${attestationUid} has already been visited`);
            }

            seen.add(attestationUid);
        },

        /**
         * Returns the set of visited attestation UIDs.
         * @returns {Set<string>} The seen set
         */
        getSeenUids() {
            return new Set(seen);
        },

        /**
         * Gets the current recursion depth.
         * @returns {number} Current depth
         */
        getDepth() {
            return currentDepth;
        },

        /**
         * Enters recursion, incrementing depth.
         * Throws if depth would exceed maxDepth.
         *
         * @throws {Error} If recursion depth would exceed maxDepth
         */
        enterRecursion() {
            if (currentDepth >= this.maxDepth) {
                throw new Error(`Recursion depth limit exceeded: max ${this.maxDepth}, current ${currentDepth}`);
            }
            currentDepth++;
        },

        /**
         * Exits recursion, decrementing depth.
         */
        exitRecursion() {
            if (currentDepth > 0) {
                currentDepth--;
            }
        },

        /**
         * Sets the validateAsync function.
         * This is called by the pipeline to allow specialists to recursively validate.
         *
         * @param {Function} fn - Async function (attestation) => AttestationResult
         */
        setValidateAsync(fn) {
            if (typeof fn !== 'function') {
                throw new Error('validateAsync must be a function');
            }
            validateAsync = fn;
        },

        /**
         * Calls the validation function recursively.
         * Must be called after setValidateAsync has been set.
         *
         * @param {Object} attestation - The attestation to validate
         * @returns {Promise<Object>} The attestation result
         * @throws {Error} If validateAsync has not been set
         */
        async validateAsync(attestation) {
            if (typeof validateAsync !== 'function') {
                throw new Error('validateAsync has not been set by the pipeline');
            }
            return validateAsync(attestation);
        }
    };
}

File: dotnet/src/Zipwire.ProofPack/ProofPack/AttestationValidationContext.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Evoq.Blockchain;

namespace Zipwire.ProofPack;

/// <summary>
/// Context for attestation validation, shared across a validation pipeline.
/// Provides cycle detection (via seen set), depth tracking, and a way for specialists
/// to recursively validate referenced attestations.
/// </summary>
public class AttestationValidationContext
{
    private readonly HashSet<string> _seenUids;
    private readonly int _maxDepth;
    private int _currentDepth;

    /// <summary>
    /// Creates a new attestation validation context.
    /// </summary>
    /// <param name="merkleRoot">Optional: The Merkle root from the document being verified.</param>
    /// <param name="maxDepth">Maximum recursion depth. Default is 32.</param>
    public AttestationValidationContext(Hex? merkleRoot = null, int maxDepth = 32)
    {
        this.MerkleRoot = merkleRoot;
        _maxDepth = maxDepth;
        _currentDepth = 0;
        _seenUids = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        this.Extension = new Dictionary<string, object>();
    }

    /// <summary>
    /// Optional Merkle root from the document being verified.
    /// Used by verifiers to validate that delegation/attestation binds to this root.
    /// </summary>
    public Hex? MerkleRoot { get; }

    /// <summary>
    /// Optional extension data for passing context-specific information.
    /// </summary>
    public IDictionary<string, object> Extension { get; }

    /// <summary>
    /// Gets the current depth in the validation chain.
    /// </summary>
    public int CurrentDepth => _currentDepth;

    /// <summary>
    /// Gets the maximum allowed depth.
    /// </summary>
    public int MaxDepth => _maxDepth;

    /// <summary>
    /// Delegate for recursive validation.
    /// Verifiers call this to validate referenced attestations (e.g., RefUID chains).
    /// </summary>
    public Func<MerklePayloadAttestation, Task<AttestationResult>>? ValidateAsync { get; set; }

    /// <summary>
    /// Records a visit to an attestation UID for cycle detection.
    /// Throws if the UID has already been visited.
    /// </summary>
    /// <param name="uid">The attestation UID to record.</param>
    /// <exception cref="InvalidOperationException">Thrown if a cycle is detected.</exception>
    public void RecordVisit(string uid)
    {
        if (string.IsNullOrEmpty(uid))
        {
            throw new ArgumentException("Attestation UID cannot be null or empty.", nameof(uid));
        }

        if (_seenUids.Contains(uid))
        {
            throw new InvalidOperationException($"Cycle detected: attestation {uid} has already been visited.");
        }

        _seenUids.Add(uid);
    }

    /// <summary>
    /// Increments recursion depth and validates it doesn't exceed max depth.
    /// Call this before recursing into ValidateAsync.
    /// </summary>
    /// <exception cref="InvalidOperationException">Thrown if max depth exceeded.</exception>
    public void EnterRecursion()
    {
        _currentDepth++;
        if (_currentDepth > _maxDepth)
        {
            throw new InvalidOperationException(
                $"Recursion depth {_currentDepth} exceeds maximum depth {_maxDepth}.");
        }
    }

    /// <summary>
    /// Decrements recursion depth.
    /// Call this after exiting a recursive ValidateAsync call.
    /// </summary>
    public void ExitRecursion()
    {
        if (_currentDepth > 0)
        {
            _currentDepth--;
        }
    }
}
